src/python/WMComponent/WorkflowUpdater/SiteListUpdater.py

```diff
@@ -0,0 +1,109 @@
+
+class SiteListUpdater(object):
+    def __init__(self, url, logger):
+        """
+        SiteListUpdater provides codebase to update site list info in local agent.
+        Its logic is based on the following:
+        - obtain list of current active workflows from the agent
+        - requests their specs from upstream ReqMgr2 server
+        - update site lists of all workflows
+        - push new specs to the agent database
+
+        :logger: logger to use
+        """
+        self.url = url
+        self.logger = logger
+
+    def getActiveRequests(self, wflowSpecs):
+        """
+        Provide list of active requests in a system (from assigned to running)
+        obtained the agent
+        wflowSpecs = self.listActiveWflows.execute()
+        :wflowSpecs: list of active workflows
+        :return: list of workflows names
+        """
+        requests = []
+        wmaSpecs = {}
+        for wflowSpec in wflowSpecs:
+            name = wflowSpec['name']
+            requests.append(name)
+            wmSpecs[name] = wflowSpec['spec']
+        # TMP: for testing purposes
+        if not requests:
+            requests = [
+                    'apiccine_ReReco_RunBlockWhite_apiccine_StageOutGfalLogs_240930_141359_1693',
+                    'apiccine_ReReco_RunBlockWhite_Nvidia_apiccine_StageOutGfalLogs_240930_141359_9090',
+                    'amaltaro_SC_MultiPU_Aug2024_Val_240730_111429_6786'
+                    ]
+        return requests, wmaSpecs
+
+    def getRequestSpecs(self, requests):
+        """
+        Return list of requests specs for provided list of request names
+        """
+        urls = []
+        for name in requests:
+            url = "{}/reqmgr2/data/request?name={}".format(self.url, name)
+            urls.append(url)
+        response = getdata(urls, ckey(), cert())
+        specs = {}
+        for resp in response:
+            data = json.loads(resp['data'])
+            for rdict in data['result']:
+                specs.update(rdict)
+        return specs
+
+    def update(self, wflowSpecs):
+        """
+        perform update logic:
+        - obtain list of current active workflows from the agent
+        - requests their specs from upstream ReqMgr2 server
+        - update site lists of all workflows
+        - push new specs to the agent database
+
+        :wflowSpecs: list of active workflows
+        :return: none
+        """
+        requests, wmaSpecs = self.getActiveRequests(wflowSpecs)
+        specs = self.getRequestSpecs(requests)
+        for key, rdict in specs.items():
+            siteList = rdict['SiteWhitelist']
+            self.logger.info(f"updating {key} site list to {siteList}")
+            print(key, rdict, siteList)
+            if key in wmaSpecs:
+                wmaSpec = wmaSpecs[key]
+                workloadHelper = WMWorkloadHelper()
+                workloadHelper.load(wmaSpec)
+                taskIterator = workloadHelper.taskIterator()
```

Comment:
no need for iterating over the object ourselves. You can simply call the relevant WMWorkload object method, e.g.:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMSpec/WMWorkload.py#L676

Reply:
yes, we do need to iterate to check if sitelist needs to be changed, otherwise we only do assignment without knowing if it is required. But I don't mind to make the assignments. The reason I structured the code to identify which task must be updated and therefore when update happens we should push back the task to whatever place it should be.